pkg/pgmodel/ingestor/ingestor.go

```diff
@@ -47,20 +49,90 @@ func NewPgxIngestorForTests(conn pgxconn.PgxConn) (*DBIngestor, error) {
 	return NewPgxIngestor(conn, c, s, &Cfg{})
 }
 
+const (
+	meta = iota
+	series
+)
+
 // Ingest transforms and ingests the timeseries data into Timescale database.
 // input:
 //     tts the []Timeseries to insert
 //     req the WriteRequest backing tts. It will be added to our WriteRequest
 //         pool when it is no longer needed.
-func (ingestor *DBIngestor) Ingest(r *prompb.WriteRequest) (uint64, error) {
+func (ingestor *DBIngestor) Ingest(r *prompb.WriteRequest) (numSamples uint64, numMetadata uint64, err error) {
 	var (
-		err       error
-		totalRows uint64
+		timeseries = r.Timeseries
+		metadata   = r.Metadata
+	)
+	// WriteRequests can contain pointers into the original buffer we deserialized
+	// them out of, and can be quite large in and of themselves. In order to prevent
+	// memory blowup, and to allow faster deserializing, we recycle the WriteRequest
+	// here, allowing it to be either garbage collected or reused for a new request.
+	// In order for this to work correctly, any data we wish to keep using (e.g.
+	// samples) must no longer be reachable from req.
+	release := func() { FinishWriteRequest(r) }
```

Comment:
There is no reason to do this kind of renaming. You should just call `FinishWriteRequest` when needed.

Reply:
Reason why `release()` is used than just calling `FinishWriteRequest` is, we need to clear mem:
1. Only insertMetadata()
2. Only insertSamples()
3. Both insertMetadata() & insertSamples()

In 1 & 2, its simple, just call the `FinishWriteRequest`. But if its 3, then wait for 1 and 2 to complete, and only then clear the request since this is done concurrently. If we hard code `FinishWriteRequest` then case 3 is not possible.

Reply:
Sorry, but I don't understand how that helps in keeping things than just overriding the function.

Reply:
I'm good with this approach with a small tweak: sending in a nil function when you can't release memory instead of an empty func and have checks for nil funcs in the ingestion functions. 

Its more clear when reading the code and adds a check that is needed anyways.